src/services/SearchService.ts

```diff
@@ -1,7 +1,281 @@
+import { Facility, Space } from '../proto/facility';
+import { DateTime, Interval } from 'luxon';
+import {
+  FacilityRuleRepository,
+  SpaceRuleRepository
+} from '../repositories/RuleRepository';
+import { DayOfWeekLOSRule, NoticeRequiredRule } from '../proto/lpms';
+import { FormattedDate, RulesItemKey } from './DBService';
+import { SpaceAvailabilityRepository } from '../repositories/SpaceAvailabilityRepository';
+import { SpaceStubRepository } from '../repositories/SpaceStubRepository';
+import ApiError from '../exceptions/ApiError';
+import { convertDaysToSeconds } from '../utils';
 
 export class SearchService {
-  public search = async (facilityId: string, ask: Ask): Promise<string[]> => {
-    throw new Error('Not Implemented Yet');
+  protected facilityRuleRepository: FacilityRuleRepository;
+  protected facilityId: string;
+  protected ask: Ask;
+  protected facilityRepository: FacilityRepository;
+  protected facility: Facility;
+  private checkInTime: string;
+
+  constructor(facilityId: string, ask: Ask) {
+    this.facilityRuleRepository = new FacilityRuleRepository(facilityId);
+    this.facilityRepository = new FacilityRepository();
+    this.facilityId = facilityId;
+    this.ask = ask;
+  }
+
+  public search = async (): Promise<string[]> => {
+    this.facility = (await this.facilityRepository.getFacilityKey(
+      this.facilityId,
+      'metadata'
+    )) as Facility;
+
+    this.checkInTime =
+      this.facility.policies?.checkInTimeOneof.oneofKind === 'checkInTime'
+        ? this.facility.policies.checkInTimeOneof.checkInTime
+        : 'utc';
+
+    const spacesIds = await facilityRepository.getFacilityKey(
+      this.facilityId,
+      'spaces'
+    );
+
+    if (Array.isArray(spacesIds)) {
+      const set = new Set();
+
+      for (const v of spacesIds) {
+        const space = await facilityRepository.getItemKey(
+          this.facilityId,
+          'spaces',
+          v,
+          'metadata'
+        );
+        set.add({ space, id: v });
+      }
+
+      return this.findSpaces(Array.from(set));
+    }
+
+    return [];
   };
+
+  private async findSpaces(spaces): Promise<string[]> {
+    const needed = new Set<string>();
+
+    const dates = this.getAskDates();
+    if (!dates.length) return [];
+
+    for (const i of spaces) {
+      const space = i.space as Space;
+
+      //check rules
+      if (!(await this.checkRules(dates, i.id))) {
+        continue;
+      }
+
+      //check space capacity
+      if (!this.checkSuitableQuantity(space)) {
+        continue;
+      }
+
+      //check dates is available
+      if (await this.checkAvailableDates(i.id, dates)) {
+        needed.add(i.id);
+      }
+    }
+
+    return Array.from(needed);
+  }
+
+  private getAskDates(): DateTime[] {
+    if (!this.ask.checkIn || !this.ask.checkOut) {
+      throw ApiError.BadRequest('invalid dates in ask');
+    }
+
+    const checkIn = DateTime.fromFormat(this.checkInTime, 'hhmm', {
```

Comment:
As this is processing _only dates_ (no times required), this should be handled just using dates, no casting to UTC or what not. Is this possible with this library?

Reply:
This is done for the convenience of further calculations, we calculate immediately in utc time,then we look at how much is left directly before the check-in